signer/xpi/cose.go

```diff
@@ -0,0 +1,339 @@
+	if err != nil {
+		err = errors.Wrapf(err, "xpi: got unexpected COSE Signature headers")
+		return
+	}
+
+	kidBytes, ok := kidValue.([]byte)
+	if !ok {
+		err = fmt.Errorf("xpi: COSE Signature kid value is not a byte array")
+		return
+	}
+
+	eeCert, err = x509.ParseCertificate(kidBytes)
+	if err != nil {
+		err = errors.Wrapf(err, "xpi: failed to parse X509 EE certificate from COSE Signature")
+		return
+	}
+
+	return
+}
+
+// isValidCOSEMessage checks whether a COSE SignMessage is a valid for
+// XPIs and returns parsed intermediate and end entity certs
+func isValidCOSEMessage(msg *cose.SignMessage) (intermediateCerts, eeCerts []*x509.Certificate, err error) {
+	if msg == nil {
+		err = errors.New("xpi: cannot validate nil COSE SignMessage")
+		return
+	}
+	if msg.Payload != nil {
+		err = fmt.Errorf("xpi: expected SignMessage payload to be nil, but got %v", msg.Payload)
+		return
+	}
+	kidValue, err := expectHeadersAndGetKeyID(msg.Headers, expectedMessageHeaders)
+	if err != nil {
+		err = errors.Wrapf(err, "xpi: got unexpected COSE SignMessage headers")
+		return
+	}
+
+	// check that all kid values are bytes and decode into certs
+	kidArray, ok := kidValue.([]interface{})
+	if !ok {
+		err = fmt.Errorf("xpi: expected SignMessage Protected Headers kid value to be an array got %v with type %T", kidValue, kidValue)
+		return
+	}
+	for i, cert := range kidArray {
+		certBytes, ok := cert.([]byte)
+		if !ok {
+			err = fmt.Errorf("xpi: expected SignMessage Protected Headers kid value %d to be a byte slice got %v with type %T", i, cert, cert)
+			return
+		}
+		intermediateCert, parseErr := x509.ParseCertificate(certBytes)
+		if parseErr != nil {
+			err = errors.Wrapf(parseErr, "xpi: SignMessage Signature Protected Headers kid value %d does not decode to a parseable X509 cert", i)
+			return
+		}
+		intermediateCerts = append(intermediateCerts, intermediateCert)
+	}
+
+	for i, sig := range msg.Signatures {
+		eeCert, sigErr := isValidCOSESignature(&sig)
+		if sigErr != nil {
+			err = errors.Wrapf(sigErr, "xpi: cose signature %d is invalid", i)
+			return
+		}
+		eeCerts = append(eeCerts, eeCert)
+	}
+
+	return
+}
+
+// verifyCOSESignatures checks that:
+//
+// 1) COSE manifest and signature files are present
+// 2) the PKCS7 manifest is present
+// 3) the COSE and PKCS7 manifests do not include COSE files
+// 4) we can decode the COSE signature and it has the right format for an XPI
+// 5) the right number of signatures are present and all intermediate and end entity certs parse properly
+// TODO: 6) there is a trusted path from the included COSE EE certs to the signer cert using the provided intermediates
+//
+func verifyCOSESignatures(signedFile signer.SignedFile, truststore *x509.CertPool, signOptions Options) error {
+	coseManifest, err := readFileFromZIP(signedFile, "META-INF/cose.manifest")
+	if err != nil {
+		return fmt.Errorf("xpi: failed to read META-INF/cose.manifest from signed zip: %v", err)
```

Comment:
we  use `errors.Wrap` in other places so maybe we should continue using it here for consistency?